fix calling id.watch() from the head

seanmonstar opened the pull request:
stores loggedInUser into a local variable, and notifies
the Channel whenever it finally builds

fixes #2252

seanmonstar committed:
fix calling id.watch() from the head

stores loggedInUser into a local variable, and notifies
the Channel whenever it finally builds

fixes #2252

shane-tomlinson commented:
@seanmonstar - I ran the PR against a modified example RP, see the attached diff.  @fmarier's observation that document.body is null is still a problem, this prevents the communication iframe from being created if navigator.id.watch is called within the head of the document.

This is not a problem in the initial RP's tab because the assertion is generated in the dialog and passed back through WinChan, making everything appear to operate correctly in the initial tab. There _is_ a problem if the user creates a new account or adds an email to an existing account and they must verify their email. When the email verification page is redirected back to the RP, navigator.id.watch tries to create the communication iframe to generate the assertion. Since document.body does not yet exist, the iframe is not created and no assertion is ever generated.

Perhaps doing a poll for the existence of document.body? Waiting for DOMContentLoaded would work in browsers that support it, for IE8 we would have to use onload. The danger of waiting for onload is that it waits for all document resources to finish, if for some reason an image takes a long time to download or times out, the user could be waiting for an extremely long time.

diff:
http://pastebin.mozilla.org/1768337

seanmonstar commented:
Oh I see what you mean now. I had thought similarly, but when I tested, I was looking at the console output. The callbacks all being triggered, since the observers are held onto regardless of the iframe. And I was clicking logout first, which tries to add the iframe again.

I'll listen for `DOMContentLoaded` and `readyStateChange` for IE8.

seanmonstar committed:
wait for domready to open iframe

if id.watch() is called in the head, and document.body
doesn't exist yet, a listener is added for domready, and then
the iframe is created.

shane-tomlinson commented on resources/static/include_js/include.js:
```diff
@@ -981,18 +983,48 @@
+          var state = document.readyState;
+          // 'interactive' is the same as DOMContentLoaded,
+          // but not all browsers use it, sadly.
+          if (['loaded', 'complete', 'interactive'].indexOf(state) > -1) {
```
IE8 unfortunately doesn't support Array.prototype.indexOf.

shane-tomlinson replied:
@seanmonstar - if I apply the following diff, IE8 acts as expected:

```
diff --git a/resources/static/include_js/include.js b/resources/static/include_js/include.js
index 979bca2..013f05f 100644
--- a/resources/static/include_js/include.js
+++ b/resources/static/include_js/include.js
@@ -997,9 +997,12 @@
           var state = document.readyState;
           // 'interactive' is the same as DOMContentLoaded,
           // but not all browsers use it, sadly.
-          if (['loaded', 'complete', 'interactive'].indexOf(state) > -1) {
-            document.detachEvent('onreadystatechange', ready);
-            callback();
+          var events = ['loaded', 'complete', 'interactive'];
+          for(var i = 0, eventName; eventName = events[i]; ++i) {
+            if(eventName === state) {
+              document.detachEvent('onreadystatechange', ready);
+              callback();
+            }
           }
         });
       }
```

seanmonstar replied:
I forget that so often...

I'm inclined to switch it to `if (state === 'loaded' || state === 'complete' || state === 'interactive')`

seanmonstar commented:
@shane-tomlinson changed away from indexOf

shane-tomlinson commented on resources/static/include_js/include.js:
```diff
@@ -981,18 +983,48 @@
+          // 'interactive' is the same as DOMContentLoaded,
+          // but not all browsers use it, sadly.
+          if (state === 'loaded' || state === 'complete' || state === 'interactive') {
+            document.detachEvent('onreadystatechange', ready);'
```
There is an extra ' on the end of this line.

shane-tomlinson commented on resources/static/include_js/include.js:
```diff
@@ -981,18 +983,48 @@
+          var state = document.readyState;
+          // 'interactive' is the same as DOMContentLoaded,
+          // but not all browsers use it, sadly.
+          if (state === 'loaded' || state === 'complete' || state === 'interactive') {
```
Much simpler than my solution.

shane-tomlinson commented:
@seanmonstar - I removed the extra ' and manually merged this in with commit 18c7a2f7c486a9f722fedf1f8d40b9cdef589cf5.  

close #2252
